### PR TITLE
nixos/searx: add configuration for redis and limiter settings

### PR DESCRIPTION
**Pull request by 999eagle**
###### Description of changes

As mentioned in #240285, this PR adds configuration options for automatically configuring a local redis instance and for configuring the new limiter in searxng.

###### Things done

- Built on platform(s)
  - [X] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [X] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: nixos/modules/services/networking/searx.nix [hunk 1]
```diff
@@ -65,6 +67,12 @@ in
         '';
       };
 
+      redisCreateLocally = mkOption {
+        type = types.bool;
+        default = true;
```
Comment (SuperSandro2000) [inline nixos/modules/services/networking/searx.nix:72]:
```suggestion
        default = false;
```

File: nixos/modules/services/networking/searx.nix [hunk 2]
```diff
@@ -206,6 +250,8 @@ in
     services.searx.settings = {
       # merge NixOS settings with defaults settings.yml
       use_default_settings = mkDefault true;
+      # set local redis URL if configured
+      redis.url = lib.mkIf cfg.redisCreateLocally "redis://localhost:6379";
```
Comment (SuperSandro2000) [inline nixos/modules/services/networking/searx.nix:254]:
```suggestion
      redis.url = lib.mkIf cfg.redisCreateLocally "unix://${config.services.redis.servers.searx.unixSocket}";
```

File: nixos/modules/services/networking/searx.nix [hunk 3]
```diff
@@ -206,6 +250,8 @@ in
     services.searx.settings = {
       # merge NixOS settings with defaults settings.yml
       use_default_settings = mkDefault true;
+      # set local redis URL if configured
```
Comment (SuperSandro2000) [inline nixos/modules/services/networking/searx.nix:253]:
```suggestion
```

File: nixos/modules/services/networking/searx.nix [hunk 4]
```diff
@@ -231,6 +277,13 @@ in
       } // cfg.uwsgiConfig;
     };
 
+    services.redis.servers.searx = lib.mkIf (cfg.redisCreateLocally) {
```
Comment (SuperSandro2000) [inline nixos/modules/services/networking/searx.nix:280]:
```suggestion
    services.redis.servers.searx = lib.mkIf cfg.redisCreateLocally {
```

File: nixos/modules/services/networking/searx.nix [hunk 5]
```diff
@@ -231,6 +277,13 @@ in
...
+      enable = true;
```
Comment (SuperSandro2000) [inline nixos/modules/services/networking/searx.nix:281]:
```suggestion
      enable = true;
      user = "searx";
```

File: nixos/modules/services/networking/searx.nix [hunk 6]
```diff
@@ -158,6 +191,17 @@ in
...
   config = mkIf cfg.enable {
+    assertions = [
+      {
+        assertion = cfg.redisCreateLocally -> (cfg.settings.redis.url == "redis://localhost:6379");
+        message = "When services.searx.redisCreateLocally is enabled, you need to use redis://localhost:6379 as a cache server.";
+      }
+      {
+        assertion = (cfg.limiterSettings != { }) -> cfg.package.pname == "searxng";
+        message = "services.searx.limiterSettings requires services.searx.package to be searxng.";
+      }
```
Comment (SuperSandro2000) [inline nixos/modules/services/networking/searx.nix:202]: Maybe we can also switch the default at some point to searxng

File: nixos/modules/services/networking/searx.nix [hunk 7]
```diff
@@ -158,6 +191,17 @@ in
   ###### implementation
 
   config = mkIf cfg.enable {
+    assertions = [
+      {
+        assertion = cfg.redisCreateLocally -> (cfg.settings.redis.url == "redis://localhost:6379");
+        message = "When services.searx.redisCreateLocally is enabled, you need to use redis://localhost:6379 as a cache server.";
+      }
```
Comment (SuperSandro2000) [inline nixos/modules/services/networking/searx.nix:198]:
```suggestion
```
I don't think we need this when using the socket

**Comment by rnhmjoj** [comment]
@999eagle Can you add yourself as maintainer? I don't use searx-ng and don't how the configuration you're adding works.

**Comment by 999eagle** [comment]
> Can you add yourself as maintainer?

Done.

The limiter settings are specific to SearXNG, but redis should work for both Searx and SearXNG.

**Review by SuperSandro2000 — APPROVED** [review]

File: nixos/modules/services/networking/searx.nix [hunk 8]
```diff
@@ -65,6 +67,12 @@ in
         '';
       };
 
+      redisCreateLocally = mkOption {
+        type = types.bool;
+        default = false;
+        description = lib.mdDoc "Configure local Redis server for searx.";
```
Comment (rnhmjoj) [inline nixos/modules/services/networking/searx.nix:73]: Is this needed for the limiter only or is also used for other purposes? In any case try to explain when it's needed. Also explain that it's a searx-ng only feature.

**Review by rnhmjoj — COMMENTED** [review]
I would also add release note since this is a useful new feature.

**Comment by rnhmjoj** [comment]
Looks good. I only suggest you move the release notes entry in a random position to avoid conflicts.

**Comment by rnhmjoj** [comment]
The test is broken, but it has been so for a while. I'm looking into it.

**Review by rnhmjoj — APPROVED** [review]
Opened #240789 to fix searx (og). The tests are passing with this branch merged, so it's good by me.

File: nixos/modules/services/networking/searx.nix (hunk 6 above)
Comment (rnhmjoj) [reply nixos/modules/services/networking/searx.nix:201]: Yeah, searx development has kind of halted and broken engines are piling up.

**Pull request merged**